Creating toString, equals, and hashCode methods in Utils class

elifKurtay opened the pull request:
Feature implementation of [this issue](https://github.com/micronaut-projects/micronaut-sourcegen/issues/126#issue-2454329573).

Main changes:
- `@ToString` annotation for classes that returns the String description of the class through the following method:
`public static String [BeanName]Utils.toString([BeanName] object)`
- `@Equals` annotation for classes that returns the result of `equals` method through the following method:
`public static boolean [BeanName]Utils.equals([BeanName] object1, Object object2)`
- `@HashCode` annotation for classes that returns the hash value of the class instance through the following method:
`public static int [BeanName]Utils.hashCode([BeanName] object)`
- `@Secret` annotation for fields that the user wishes not to print in the toString function. Instead of printing the value, this annotation will print a fixed `"******"` representation. 

For `Person4` class in `test-suite-java` the generated Utils are:
```java
package io.micronaut.sourcegen.example;

import java.lang.Object;
import java.lang.String;
import java.lang.StringBuilder;
import java.util.Arrays;

public final class Person4Utils {
  public static String toString(Person4 object) {
    StringBuilder strBuilder = new java.lang.StringBuilder();
    strBuilder.append("Person4[");
    strBuilder.append("id=").append(object.getId()).append(", ");
    strBuilder.append("title=").append(object.getTitle()).append(", ");
    strBuilder.append("name=").append(object.getName()).append(", ");
    strBuilder.append("bytes=").append(java.util.Arrays.toString(object.getBytes())).append("]");
    return strBuilder.toString();
  }

  public static boolean equals(Person4 first, Object secondObject) {
    if (first == secondObject) {
      return true;
    }
    boolean isCorrectInstance = secondObject instanceof Person4;
    if (isCorrectInstance == false) {
      return false;
    }
    Person4 second = (Person4) (secondObject);
    boolean bothNullCondition = false;
    boolean equalsCondition = false;
    boolean isPropertyEqual = false;
    if (first.getId() != second.getId()) {
      return false;
    }
    bothNullCondition = first.getTitle() == null && second.getTitle() == null;
    equalsCondition = first.getTitle() != null && first.getTitle().equals(second.getTitle());
    isPropertyEqual = bothNullCondition || equalsCondition;
    if (isPropertyEqual == false) {
      return false;
    }
    bothNullCondition = first.getName() == null && second.getName() == null;
    equalsCondition = first.getName() != null && first.getName().equals(second.getName());
    isPropertyEqual = bothNullCondition || equalsCondition;
    if (isPropertyEqual == false) {
      return false;
    }
    bothNullCondition = first.getBytes() == null && second.getBytes() == null;
    equalsCondition = first.getBytes() != null && Arrays.equals(first.getBytes(), second.getBytes());
    isPropertyEqual = bothNullCondition || equalsCondition;
    if (isPropertyEqual == false) {
      return false;
    }
    return true;
  }

  public static int hashCode(Person4 object) {
    int hashValue = 1;
    hashValue = hashValue * 59 + (int) (object.getId() >>> 32 ^ object.getId());
    hashValue = hashValue * 59 + (int) (object.getTitle() == null ? 43 : object.getTitle().hashCode());
    hashValue = hashValue * 59 + (int) (object.getName() == null ? 43 : object.getName().hashCode());
    hashValue = hashValue * 59 + (int) (java.util.Arrays.hashCode(object.getBytes()));
    return hashValue;
  }
}
```

elifKurtay committed:
utils annotation with tostring, equals, and hashcode functions together

elifKurtay committed:
Merge remote-tracking branch 'origin/1.3.x' into 1.3.x

elifKurtay committed:
code formatting and cast modify

elifKurtay committed:
change to static functions

elifKurtay committed:
fix basic tests

elifKurtay committed:
create separate annotations

elifKurtay committed:
updated tests

elifKurtay committed:
add @Secret for @ToString

elifKurtay committed:
clean up

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/micronaut-projects/micronaut-sourcegen?pullRequest=149) <br/>All committers have signed the CLA.

andriy-dmytruk commented on sourcegen-generator-java/src/main/java/io/micronaut/sourcegen/JavaPoetSourceGenerator.java:
```diff
@@ -538,8 +538,9 @@ private CodeBlock renderExpression(@Nullable ObjectDef objectDef, MethodDef meth
-                renderExpression(objectDef, methodDef, castExpressionDef.expressionDef())
+                CodeBlock.of("($T) (", asType(castExpressionDef.type(), objectDef)),
+                renderExpression(objectDef, methodDef, castExpressionDef.expressionDef()),
+                CodeBlock.of(")")
```
As Elif mentioned to me, it is hard to change order of operations in expressions and maybe a way of adding brackets would be good.

dstepanov replied:
You can probably check if the expression is an instance of a variable and exclude the brackets

dstepanov commented:
Maybe instead of `MyBeanUtils` we can call it `MyBeanObject` or `MyBeanCompanion`.

```java
 public static String toString(Person4 object) {
    StringBuilder strBuilder = new java.lang.StringBuilder();
    strBuilder.append("Person4[");
    strBuilder.append("id=").append(object.getId()).append(", ");
    strBuilder.append("title=").append(object.getTitle()).append(", ");
    strBuilder.append("name=").append(object.getName()).append(", ");
    strBuilder.append("bytes=").append(java.util.Arrays.toString(object.getBytes())).append("]");
    return strBuilder.toString();
  }
```
You can add the first line in the constructor:
```java
    StringBuilder strBuilder = new java.lang.StringBuilder("Person4[");
```

```java
public static boolean equals(Person4 first, Object secondObject) {
    if (first == secondObject) {
      return true;
    }
    boolean isCorrectInstance = secondObject instanceof Person4;
    if (isCorrectInstance == false) {
      return false;
    }
    Person4 second = (Person4) (secondObject);
    boolean bothNullCondition = false;
    boolean equalsCondition = false;
    boolean isPropertyEqual = false;
    if (first.getId() != second.getId()) {
      return false;
    }
    bothNullCondition = first.getTitle() == null && second.getTitle() == null;
    equalsCondition = first.getTitle() != null && first.getTitle().equals(second.getTitle());
    isPropertyEqual = bothNullCondition || equalsCondition;
    if (isPropertyEqual == false) {
      return false;
    }
    bothNullCondition = first.getName() == null && second.getName() == null;
    equalsCondition = first.getName() != null && first.getName().equals(second.getName());
    isPropertyEqual = bothNullCondition || equalsCondition;
    if (isPropertyEqual == false) {
      return false;
    }
    bothNullCondition = first.getBytes() == null && second.getBytes() == null;
    equalsCondition = first.getBytes() != null && Arrays.equals(first.getBytes(), second.getBytes());
    isPropertyEqual = bothNullCondition || equalsCondition;
    if (isPropertyEqual == false) {
      return false;
    }
    return true;
  }
```

Please generate the code more or less like IntellJ equals generator, and avoid extra variables. Please also use `getClass` to compare classes. HashCode should have similar code like intellij:

```java
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Person4 person4 = (Person4) o;
        return id == person4.id && title == person4.title && Objects.equals(name, person4.name) && Objects.deepEquals(bytes, person4.bytes);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, title, name, Arrays.hashCode(bytes));
    }
```

dstepanov commented on sourcegen-annotations/src/main/java/io/micronaut/sourcegen/annotations/Equals.java:
```diff
@@ -0,0 +1,34 @@
+
+@Retention(RUNTIME)
+@Target({ElementType.ANNOTATION_TYPE, ElementType.TYPE})
+public @interface Equals {
```
There is no point to have two annotation. Please have only one `EqualsAndHashCode`

dstepanov commented on sourcegen-annotations/src/main/java/io/micronaut/sourcegen/annotations/Secret.java:
```diff
@@ -0,0 +1,36 @@
+
+@Retention(RUNTIME)
+@Target({FIELD, PARAMETER})
+public @interface Secret {
```
I would prefer to have the same annotation as Lombok has (Exclude, Include) and have them as an inner annotation of `ToString` so you know where does it belong: `@ToString.Include`

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+            .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+            .returns(String.class)
+            .addParameter("object", ClassTypeDef.of(objectName))
+            .build((self, parameterDef) -> {
```
Rewrite this body as:

```java
ClassTypeDef.of(StringBuilder.class).instantiate().newLocal("strBuilder", variableDef -> {
      ExpressionDef exp = variableDef;
      for (PropertyElement propertyElement : properties) {
          exp = exp.invoke("append", variableDef.type(), ...);
      }
      return exp.invoke("toString", TypeDef.of(String.class)).returning();
});
```

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+                        ClassTypeDef.of(strBuilder.getClass()),
+                        (TypeDef.of(beanProperty.getType()) instanceof TypeDef.Array) ?
+                            new ExpressionDef.CallStaticMethod(
+                                ClassTypeDef.of(Arrays.class),
```
Please fluent API:
`ClassTypeDef.of(Arrays.class).invokeStatic(...)`

andriy-dmytruk replied:
We should probably improve the docs by providing examples of how to do most common operations. Something like:

| Operation | Example code | Genereated example |
| --- | --- | --- |
| ... | ... | ... |
| Static invocation | `ClassTypeDef.of(String.class).invokeStatic("valueOf", ClassTypeDef.of(String.class), ExpressionDef.constant(2))` | `String.valueOf(2)` |

graemerocher replied:
I think part of this PR is as we add more functionality we should extend the API of SourceGen to improve usability at the same time. It is an evolving API and clearly from the looks of this PR could do with enhancements to make writing code that writes code simpler.

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+            .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+            .addParameter("object", ClassTypeDef.of(objectName))
+            .returns(int.class)
+            .build((self, parameterDef) -> {
```
You should be able to write this as:

```java
  .build((self, parameterDef) -> {
      VariableDef objectVar = parameterDef.get(0).asVariable();
      return StatementDef.multi(
          objectVar.asCondition(" == ", self)
              .asConditionIf(ExpressionDef.trueValue().returning()),
          // TODO: if (o == null || getClass() != o.getClass()) {
          //            return false;
          //        }
          objectVar.cast(self.type()).newLocal("instance", variableDef -> {
              ExpressionDef exp = null;
              for (PropertyElement property : properties) {
                  ExpressionDef newEqualsExpression = ...;
                  if (exp == null) {
                      exp = newEqualsExpression;
                  } else {
                      exp = exp.asCondition(" && ", newEqualsExpression);
                  }
              }
              return exp.returning();
          })
  ));
```

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+                            propertyHashExpression = thisProperty;
+                        }
+                    } else { // OBJECT
+                        propertyHashExpression = new ExpressionDef.IfElse(
```
Please try to avoid using `new ExpressionDef` but use a a fluent API

dstepanov commented:
Please sign the License Agreement and add docs.

dstepanov commented:
We should consider to add as much as possible of what Lombok supports (can be added in another PR):

- https://projectlombok.org/api/lombok/EqualsAndHashCode
- https://projectlombok.org/api/lombok/ToString 

We should support `exclude`, `include`

The `callSuper` might be needed as well, in the case of false we should only include the properties in the current class

graemerocher commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+ */
+
+@Internal
+public final class UtilsAnnotationVisitor implements TypeElementVisitor<Object, Object> {
```
It feels like this implementation is doing too much. Ideally it would be split into separate visitors and add the annotation type to the generic arguments. So `EqualsAndHashCodeVisitor implements TypeElementVisitor<EqualsHashCode, Object>`, `ToStringVisitor implements TypeElementVisitor<ToString, Object>` etc. however I am aware that will require generating multiple Utils classes so maybe not ideal

Either way you need to override `getSupportedAnnotationNames()` and list the annotations this visitor processes.

graemerocher commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+                VariableDef firstObject = parameterDef.get(0).asVariable();
+                VariableDef.Local secondObject = new VariableDef.Local("second", classname.type());
+                VariableDef.Local bothNullCondition = new VariableDef.Local("bothNullCondition", TypeDef.of(boolean.class));
+                VariableDef.Local equalsCondition = new VariableDef.Local("equalsCondition", TypeDef.of(boolean.class));
```
use `var` on the left hand side of the statement to make these easier to read

graemerocher commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+                        propertyHashExpression = arrays.invoke(methodName, TypeDef.of(int.class), thisProperty);
+                    } else if (propertyTypeDef instanceof TypeDef.Primitive) {
+                        String typeName = ((TypeDef.Primitive) propertyTypeDef).name();
+                        if (propertyTypeDef == TypeDef.BOOLEAN) {
```
rewrite these nested `ifs` as a switch expression that assigns `propertyHashExpression`

graemerocher commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+                ExpressionDef thisProperty;
+                for (PropertyElement beanProperty : properties) {
+                    propertyTypeDef = TypeDef.of(beanProperty.getType());
+                    if (beanProperty.getReadMethod().isPresent()) {
```
extract `beanProperty.getReadMethod()` into a local variable and reuse

graemerocher commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+                    } else {
+                        // .equals, check for double null or equal objects
+                        statements.add(new StatementDef.Assign(bothNullCondition,
+                                new ExpressionDef.Condition(" && ",
```
can we upstream improve the `ExpressionDef.Condition` type to make it easier to define `&&` expressions? //cc @dstepanov

dstepanov replied:
All this should be written using the fluent API, for some basic conditions we can maybe add `and`/`or`

elifKurtay replied:
Should I try to do a basic implementation now or is this a thought for future? 

In my opinion, integrating parentheses into expressions and/or creating a `not` representation (`!(expression)`) would be more useful.

graemerocher commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+                            statements.add(new StatementDef.Assign(equalsCondition,
+                                    new ExpressionDef.Condition(" && ",
+                                        firstProperty.isNonNull(),
+                                        new ExpressionDef.CallStaticMethod(
```
Use `ExpressionDef.invokeStatic` instead

graemerocher commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+                List<StatementDef> statements = new ArrayList<>();
+                ExpressionDef propertyHashExpression;
+                VariableDef.Local hashValue = new VariableDef.Local("hashValue", TypeDef.of(int.class));
+                VariableDef arrays = new VariableDef.Local("java.util.Arrays", ClassTypeDef.of(java.util.Arrays.class));
```
this seems an odd variable name, looks like a type name?

graemerocher commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java:
```diff
@@ -0,0 +1,406 @@
+                                ExpressionDef.constant(FALSE_HASH_VALUE)
+                            );
+                        } else if (typeName.equals("float")) {
+                            VariableDef floatClass = new VariableDef.Local("Float", ClassTypeDef.of(float.class));
```
again non-consistent variable nam used. Also consider using constants for names

graemerocher reviewed (changes requested):
Great start! Well done @elifKurtay

elifKurtay committed:
-variable check at casting
-Utils -> Object
-simplify equals function

elifKurtay committed:
-simplify hashcode
-merge equals and hashcode
-getSupportedAnnotationNames
-fluent api

dstepanov commented:
Please try to refactor it to look like my suggestions

andriy-dmytruk commented:
@dstepanov I don't think that `Objects.hash` is optimal. For primitives, it needs to wrap them to a class and then call `Object.hashCode`. Not that it is significant, but could be important, since it can be called frequently in collections. Lombok generates custom code for these types. The initial implementation was based on that.

Similarly for equals, using `==` seems like a very simple check, but could be slightly more optimal.

andriy-dmytruk commented:
> The callSuper might be needed as well, in the case of false we should only include the properties in the current class

We thought about this, but unfortunately we cannot call super from a different class. Generating methods using parent properties is possible, but not sure if that is desirable, as there could be custom behavior in the parent.

dstepanov commented:
I'm OK with inlining `Objects.equals` and `Objects.hash` which is:
```
    public static int hashCode(Object a[]) {
        if (a == null)
            return 0;

        int result = 1;

        for (Object element : a)
            result = 31 * result + (element == null ? 0 : element.hashCode());

        return result;
    }
```
I don't think we should copy exactly what Lobmok does without a reason.

> We thought about this, but unfortunately we cannot call super from a different class. Generating methods using parent properties is possible, but not sure if that is desirable, as there could be custom behavior in the parent.

That's not about calling the superclass; that's about including/excluding properties from the superclass.

elifKurtay commented:
> Please try to refactor it to look like my suggestions

I was still improving :') I will re-request review once I finished addressing all comments and requests 👍

elifKurtay committed:
-replace secret with Exclude
-refactor code to be more fluent
-rollback builder

elifKurtay commented:
@andriy-dmytruk @dstepanov What is the verdict on `equals` and `hashCode` functions?

I think the current implementation of `equals` is good since it uses `==` for primitives and `Objects.equals` for objects. And apparently, `Objects.equals` check for both null before calling `.equals` on objects and `Array.equals` on arrays which is exactly equal to our initial implementation.

Should I go back to the initial description [here](https://github.com/micronaut-projects/micronaut-sourcegen/issues/126#issuecomment-2363808621) for `hashCode`? In addition, I think we can call the super class like:

```
public static int hashCode(Object object) {
        if (object == null)
            return 0;

        int result = object.getClass().getSuperclass().hashCode();
        
        result = 31 * result + (object.getElement() == null ? 0 : object.getElement().hashCode());
        ...

        return result;
    }
```

dstepanov commented:
I think you can inline `Objects.equals` which is `(a == b) || (a != null && a.equals(b))` and hashCode too

dstepanov commented:
`object.getClass().getSuperclass().hashCode()` this doesn't look correct

elifKurtay commented:
> We should consider to add as much as possible of what Lombok supports (can be added in another PR):
> 
> * https://projectlombok.org/api/lombok/EqualsAndHashCode
> * https://projectlombok.org/api/lombok/ToString
> 
> We should support `exclude`, `include`
> 
> The `callSuper` might be needed as well, in the case of false we should only include the properties in the current class

I am adding `Exclude` for both annotations but I think there are some things to consider when it comes to `Include`.
1. As we created a static method in a separate class, we are currently only using bean properties (a.k.a. properties with getters). We have no way of accessing methods or fields that are private and without a getter even if users put the `Include` annotation. So, I think we need to create a warning system for users otherwise it will seem like a bug. I can modify the implementation to check for public fields and methods but that's about it.
2. In Lombok, `Include` has two parameters: `name` and `rank`. Name is used to override the name displayed for the field/method. Rank is to modify processing order. This can be done for `ToString`, but it seems a bit useless in `EqualsAndHashCode` to me. Should I include these parameters or not?

For `callSuper`, there is again the only option of `object.getClass().getSuperclass()` which has a [`getDeclaredFields()`](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getDeclaredFields--) function. Though I never tried these methods, I can experiment. :)

Please let me know what do you think @dstepanov @graemerocher @andriy-dmytruk?

dstepanov commented:
> For callSuper, there is again the only option of object.getClass().getSuperclass() which has a [getDeclaredFields()](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getDeclaredFields--) function. Though I never tried these methods, I can experiment. :)

That's something else, we would need to fetch properties that are only declared on the current bean. We can do it later.

elifKurtay commented:
> I think you can inline `Objects.equals` which is `(a == b) || (a != null && a.equals(b))` and hashCode too

This would require two/three variables since we don't have parentheses. Or I can "hack" it by casting the two sides to boolean for no reason 😄

dstepanov commented:
What do you mean by `we don't have parentheses`?

elifKurtay committed:
Exclude and inline hashCode

elifKurtay commented:
> What do you mean by `we don't have parentheses`?

Is there an ExpressionDef that adds parentheses around an expression like: `( <expression> )`?
asCondition does not add them as far as I tried. But maybe I am missing something?

dstepanov commented:
I see. You can add parentheses by default or introduce AND / OR conditions where you can add parentheses only for OR

elifKurtay committed:
remove Include

elifKurtay committed:
and/or expressions added

elifKurtay committed:
inline equals and parentheses by default over and/or conditions

elifKurtay committed:
clean docs

graemerocher commented on sourcegen-annotations/src/main/java/io/micronaut/sourcegen/annotations/ToString.java:
```diff
@@ -17,8 +17,7 @@
 
-import static java.lang.annotation.ElementType.FIELD;
-import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.ElementType.*;
```
avoid star imports

graemerocher commented on sourcegen-generator-java/src/main/java/io/micronaut/sourcegen/JavaPoetSourceGenerator.java:
```diff
@@ -54,10 +54,7 @@
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
configure your IDE to avoid automatic star imports

elifKurtay commented:
Lastest changes:
- Exclude for ToString and EqualsAndHashCode (no Include for the time being.)
- And/Or Expressions along with asConditionAnd/ asConditionOr that adds parentheses to its expressions when the expression is a type of ExpressionDef.Condition/ExpressionDef.And/ExpressionDef.Or.
- Inline implementations of equals and hashCode

An example Person4.java object:
```
@ToString
@EqualsAndHashCode
public class Person4 {
    public enum Title {
        MRS,
        MR,
        MS
    }

    private long id;
    private Title title;
    private String name;
    private byte[] bytes;

    public Person4(long id, Title title, String name, byte[] bytes) {
        this.id = id;
        this.title = title;
        this.name = name;
        this.bytes = (bytes != null) ? Arrays.copyOf(bytes, bytes.length) : null;
    }

    public long getId() {
        return id;
    }

    public Title getTitle() {
        return title;
    }

    public String getName() {
        return name;
    }

    public byte[] getBytes() {
        return bytes;
    }

    @Override
    public String toString() {
        return Person4Object.toString(this);
    }

    @Override
    public boolean equals(Object o) {
        return Person4Object.equals(this, o);
    }

    @Override
    public int hashCode() {
        return Person4Object.hashCode(this);
    }
}
```

And its generated Person4Object class:
```
public final class Person4Object {
  public static String toString(Person4 instance) {
    StringBuilder strBuilder = new java.lang.StringBuilder("Person4[");
    return strBuilder.append("id=").append(instance.getId()).append(", ").append("title=").append(instance.getTitle()).append(", ").append("name=").append(instance.getName()).append(", ").append("bytes=").append(java.util.Arrays.toString(instance.getBytes())).append("]").toString();
  }

  public static boolean equals(Person4 instance, Object o) {
    if (instance == o) {
      return true;
    }
    if (o == null || instance.getClass() != o.getClass()) {
      return false;
    }
    Person4 other = (Person4) o;
    return (((instance.getId() == other.getId()) && ((instance.getTitle() == other.getTitle()) || ((instance.getTitle() != null) && instance.getTitle().equals(other.getTitle())))) && ((instance.getName() == other.getName()) || ((instance.getName() != null) && instance.getName().equals(other.getName())))) && ((instance.getBytes() == other.getBytes()) || ((instance.getBytes() != null) && Arrays.equals(instance.getBytes(), other.getBytes())));
  }

  public static int hashCode(Person4 instance) {
    if (instance == null) {
      return 0;
    }
    int hashValue = 1;
    hashValue = hashValue * 59 + (int) (instance.getId() >>> 32 ^ instance.getId());
    hashValue = hashValue * 59 + (int) (instance.getTitle() == null ? 43 : instance.getTitle().hashCode());
    hashValue = hashValue * 59 + (int) (instance.getName() == null ? 43 : instance.getName().hashCode());
    hashValue = hashValue * 59 + (int) (Arrays.hashCode(instance.getBytes()));
    return hashValue;
  }
}
```

graemerocher commented on sourcegen-generator-java/src/main/java/io/micronaut/sourcegen/JavaPoetSourceGenerator.java:
```diff
@@ -581,6 +578,36 @@ private CodeBlock renderExpression(@Nullable ObjectDef objectDef, MethodDef meth
+                (isRightCondition) ? CodeBlock.of("(") : CodeBlock.of(""),
+                renderExpression(objectDef, methodDef, andExpressionDef.right()),
+                (isRightCondition) ? CodeBlock.of(")") : CodeBlock.of("")
+            );
```
split this into multiple `if/else` to make this easier to read and/or consider rewriting to use switch expressions that assign the code block

elifKurtay replied:
Wouldn't that be too many repeated lines? I am cool to change just thought it was too long for such small changes. 😄  

```
if (leftCondition && rightCondition) {
return CodeBlock.concat(
                CodeBlock.of("("),
                renderExpression(objectDef, methodDef, andExpressionDef.left()),
                CodeBlock.of(")"),
                CodeBlock.of(" && "),
                CodeBlock.of("("),
                renderExpression(objectDef, methodDef, andExpressionDef.right()),
                CodeBlock.of(")")
            );
} else if ( rightCondition) {
return CodeBlock.concat(
                renderExpression(objectDef, methodDef, andExpressionDef.left()),
                CodeBlock.of(" && "),
                CodeBlock.of("("),
                renderExpression(objectDef, methodDef, andExpressionDef.right()),
                CodeBlock.of(")")
            );
} else if (leftCondition) {
return CodeBlock.concat(
                CodeBlock.of("("),
                renderExpression(objectDef, methodDef, andExpressionDef.left()),
                CodeBlock.of(")"),
                CodeBlock.of(" && "),
                renderExpression(objectDef, methodDef, andExpressionDef.right()),
            );
} else {
return CodeBlock.concat(
                renderExpression(objectDef, methodDef, andExpressionDef.left()),
                CodeBlock.of(" && "),
                renderExpression(objectDef, methodDef, andExpressionDef.right()),
            );
} 
```

graemerocher commented on sourcegen-generator-java/src/main/java/io/micronaut/sourcegen/JavaPoetSourceGenerator.java:
```diff
@@ -581,6 +578,36 @@ private CodeBlock renderExpression(@Nullable ObjectDef objectDef, MethodDef meth
+                CodeBlock.of(" || "),
+                (isRightCondition) ? CodeBlock.of("(") : CodeBlock.of(""),
+                renderExpression(objectDef, methodDef, orExpressionDef.right()),
+                (isRightCondition) ? CodeBlock.of(")") : CodeBlock.of("")
```
split this into multiple `if/else` to make this easier to read and/or consider rewriting to use switch expressions that assign the code block

elifKurtay committed:
star import fix

elifKurtay committed:
add primitive "initialize" instead of  instantiate

elifKurtay committed:
added isPrimitive and isArray to TypeDef

elifKurtay committed:
added comment

dstepanov commented on sourcegen-generator-java/src/main/java/io/micronaut/sourcegen/JavaPoetSourceGenerator.java:
```diff
@@ -583,15 +589,39 @@ private CodeBlock renderExpression(@Nullable ObjectDef objectDef, MethodDef meth
-                renderExpression(objectDef, methodDef, andExpressionDef.right()),
-                (isRightCondition) ? CodeBlock.of(")") : CodeBlock.of("")
-            );
+            if (isLeftCondition && isRightCondition) {
```
This looks too complicated, can you just have List and add codeblock that are needed then call concat

elifKurtay replied:
Like my previous implementation?
```
if (expressionDef instanceof ExpressionDef.And andExpressionDef) {
            boolean isLeftCondition = andExpressionDef.left() instanceof ExpressionDef.Condition ||
                andExpressionDef.left() instanceof ExpressionDef.And || andExpressionDef.left() instanceof ExpressionDef.Or;
            boolean isRightCondition = andExpressionDef.right() instanceof ExpressionDef.Condition ||
                andExpressionDef.right() instanceof ExpressionDef.And || andExpressionDef.right() instanceof ExpressionDef.Or;
            return CodeBlock.concat(
                (isLeftCondition) ? CodeBlock.of("(") : CodeBlock.of(""),
                renderExpression(objectDef, methodDef, andExpressionDef.left()),
                (isLeftCondition) ? CodeBlock.of(")") : CodeBlock.of(""),
                CodeBlock.of(" && "),
                (isRightCondition) ? CodeBlock.of("(") : CodeBlock.of(""),
                renderExpression(objectDef, methodDef, andExpressionDef.right()),
                (isRightCondition) ? CodeBlock.of(")") : CodeBlock.of("")
            );
 }
```
I changed this according to this review: https://github.com/micronaut-projects/micronaut-sourcegen/pull/149#discussion_r1778742157

dstepanov replied:
Something like this:
```
        List<CodeBlock> codeBlocks = new ArrayList<>();
            if (isLeftCondition) {
                codeBlocks.add(CodeBlock.of("("));
                codeBlocks.add(renderExpression(objectDef, methodDef, andExpressionDef.left()));
                codeBlocks.add(CodeBlock.of(")"));
            } else {
                codeBlocks.add(renderExpression(objectDef, methodDef, andExpressionDef.left());
            }
```
you can also create a new method to reduce the duplicity of checking for left and right

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -234,67 +234,75 @@ private static void createHashCodeMethod(ClassDef.ClassDefBuilder classDefBuilde
+            propertyHashCalculation = ClassTypeDef.of(Arrays.class).invokeStatic(methodName, TypeDef.of(int.class), thisProperty);
+        } else if (propertyTypeDef.isPrimitive()) {
+            String typeName = ((TypeDef.Primitive) propertyTypeDef).name();
+            if (propertyTypeDef == TypeDef.BOOLEAN) {
```
Use equals methods

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -234,67 +234,75 @@ private static void createHashCodeMethod(ClassDef.ClassDefBuilder classDefBuilde
+                    ExpressionDef.constant(TRUE_HASH_VALUE),
+                    ExpressionDef.constant(FALSE_HASH_VALUE)
+                );
+            } else if (typeName.equals("float")) {
```
Please add this primitives as constants to `TypeDef.Primitive` and move the TypeDef.BOOLEAN there, so we have `TypeDef.Primitive.BOOLEAN`

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -234,67 +234,75 @@ private static void createHashCodeMethod(ClassDef.ClassDefBuilder classDefBuilde
+                    ExpressionDef.constant(FALSE_HASH_VALUE)
+                );
+            } else if (typeName.equals("float")) {
+                propertyHashCalculation = ClassTypeDef.of(Float.class).invokeStatic("floatToIntBits", TypeDef.of(int.class), thisProperty);
```
`TypeDef.of(int.class)` should be `TypeDef.Primitive.INT`

dstepanov commented on sourcegen-model/src/main/java/io/micronaut/sourcegen/model/ExpressionDef.java:
```diff
@@ -481,6 +494,19 @@ record NewInstance(ClassTypeDef type,
+     * @since 1.0
+     */
+    @Experimental
+    record PrimitiveInstance(TypeDef type,
```
Please move this to `TypeDef.Primitive`, so we call `TypeDef.Primitive.INT.instantiate(1)`

dstepanov commented on sourcegen-model/src/main/java/io/micronaut/sourcegen/model/TypeDef.java:
```diff
@@ -249,13 +249,37 @@ default boolean isNullable() {
+     * @return Is primitive type
+     */
+    default boolean isPrimitive() {
+        return this instanceof TypeDef.Primitive;
```
Override this new two methods

dstepanov commented on sourcegen-model/src/main/java/io/micronaut/sourcegen/model/TypeDef.java:
```diff
@@ -249,13 +249,37 @@ default boolean isNullable() {
+     * @param value The initial value
+     * @return The instantiate expression
+     */
+    default ExpressionDef initialize(ExpressionDef value) {
```
This doesn't make sense considering TypeDef can be `TypeDef.Wildcard`

dstepanov commented on sourcegen-model/src/main/java/io/micronaut/sourcegen/model/ExpressionDef.java:
```diff
@@ -569,6 +614,40 @@ public TypeDef type() {
+    record And(ExpressionDef left, ExpressionDef right) implements ExpressionDef {
+        @Override
+        public TypeDef type()  {
+            return TypeDef.of(boolean.class);
```
Can be a constant

dstepanov commented on sourcegen-model/src/main/java/io/micronaut/sourcegen/model/ExpressionDef.java:
```diff
@@ -569,6 +614,40 @@ public TypeDef type() {
+    record Or(ExpressionDef left, ExpressionDef right) implements ExpressionDef {
+        @Override
+        public TypeDef type()  {
+            return TypeDef.of(boolean.class);
```
Can be a constant

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -0,0 +1,311 @@
+                createToStringMethod(objectBuilder, element.getSimpleName(), properties);
+            }
+            if (element.hasStereotype(EqualsAndHashCode.class)) {
+                createEqualsMethod(objectBuilder, element.getSimpleName(), properties);
```
Please pass `ClassElement` and create a type from it

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -0,0 +1,311 @@
+                            .invoke("append", variableDef.type(),
+                                (TypeDef.of(beanProperty.getType()).isArray()) ?
+                                    ClassTypeDef.of(Arrays.class).invokeStatic("toString",
+                                        TypeDef.of(String.class), List.of(propertyValue)) : propertyValue)
```
Create a constant for STRING

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -0,0 +1,311 @@
+            .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+            .returns(boolean.class)
+            .addParameter("instance", ClassTypeDef.of(objectName))
+            .addParameter("o", TypeDef.of(Object.class))
```
Use constant for Object

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -0,0 +1,311 @@
+    Creates an equals method with signature:
+        public static boolean BeanNameObject.equals(BeanName object1, Object object2)
+     */
+    private static void createEqualsMethod(ClassDef.ClassDefBuilder classDefBuilder, String objectName, List<PropertyElement> properties) {
```
Replace `objectName` with `ClassTypeDef selfType`

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -0,0 +1,311 @@
+                                continue;
+                            }
+                            ExpressionDef newEqualsExpression = propertyTypeDef.isPrimitive() ?
+                                firstProperty.asCondition(" == ", secondProperty)
```
Duplicate code

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -0,0 +1,311 @@
+                                : firstProperty.asCondition(" == ", secondProperty)
+                                .asConditionOr(firstProperty.isNonNull().asConditionAnd(
+                                    (propertyTypeDef.isArray()) ?
+                                        ClassTypeDef.of(Arrays.class).invokeStatic("equals", TypeDef.BOOLEAN, Arrays.asList(firstProperty, secondProperty))
```
Why is it a list of values and not just `"equals", TypeDef.BOOLEAN, firstProperty, secondProperty`?

elifKurtay replied:
Because there is support for both and (ExpressionDef... parameters) gets cast into a List anyway. But I can change

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -0,0 +1,311 @@
+                            .asCondition(" != ", o.invoke("getClass", ClassTypeDef.of("Class"))))
+                        .asConditionIf(ExpressionDef.falseValue().returning()),
+                    o.cast(ClassTypeDef.of(objectName)).newLocal("other", variableDef -> {
+                        ExpressionDef exp = null;
```
Avoid variable definitions on the top

elifKurtay replied:
But this increases readability. I thought that is also important?

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -0,0 +1,311 @@
+            .build((self, parameterDef) -> StatementDef.multi(
+                    parameterDef.get(0).asExpression().isNull().asConditionIf(ExpressionDef.constant(0).returning()),
+                    TypeDef.of(int.class).initialize(ExpressionDef.constant(1)).newLocal("hashValue", hashValue -> {
+                        List<StatementDef> hashUpdates = new ArrayList<>();
```
Avoid variable definitions on the top

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -0,0 +1,311 @@
+
+                        for (PropertyElement beanProperty : properties) {
+                            propertyTypeDef = TypeDef.of(beanProperty.getType());
+                            if (!beanProperty.hasAnnotation(EqualsAndHashCode.Exclude.class) && beanProperty.getReadMethod().isPresent()) {
```
You keep repeating this kind of code everywhere, just check the value and continue; next block initialize a variable and get the value

dstepanov commented on sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java:
```diff
@@ -0,0 +1,311 @@
+            String typeName = ((TypeDef.Primitive) propertyTypeDef).name();
+            if (propertyTypeDef == TypeDef.BOOLEAN) {
+                propertyHashCalculation = property.asConditionIfElse(
+                    ExpressionDef.constant(TRUE_HASH_VALUE),
```
Use conditions

elifKurtay committed:
add new test for multidimentional arrays

elifKurtay committed:
fix and/or

elifKurtay committed:
modify primitive types

elifKurtay committed:
import problem

elifKurtay committed:
update docs
